client/executor/wasmtime/src/host.rs

```diff
@@ -25,40 +25,30 @@ use log::trace;
 use sc_allocator::FreeingBumpHeapAllocator;
 use sc_executor_common::{
 	error::Result,
-	sandbox::{self, SandboxCapabilities, SandboxCapabilitiesHolder, SupervisorFuncIndex},
+	sandbox::{self, SupervisorFuncIndex},
 	util::MemoryTransfer,
 };
 use sp_core::sandbox as sandbox_primitives;
 use sp_wasm_interface::{FunctionContext, MemoryId, Pointer, Sandbox, WordSize};
-use std::{cell::RefCell, rc::Rc};
+use std::{cell::RefCell, collections::HashMap, rc::Rc};
 use wasmtime::{Func, Val};
 
-/// Wrapper type for pointer to a Wasm table entry.
-///
-/// The wrapper type is used to ensure that the function reference is valid as it must be unsafely
-/// dereferenced from within the safe method `<HostContext as SandboxCapabilities>::invoke`.
-#[derive(Clone)]
-pub struct SupervisorFuncRef(Func);
-
 /// The state required to construct a HostContext context. The context only lasts for one host
 /// call, whereas the state is maintained for the duration of a Wasm runtime call, which may make
 /// many different host calls that must share state.
-#[derive(Clone)]
 pub struct HostState {
-	inner: Rc<Inner>,
-}
-
-struct Inner {
-	// We need some interior mutability here since the host state is shared between all host
-	// function handlers and the wasmtime backend's `impl WasmRuntime`.
-	//
-	// Furthermore, because of recursive calls (e.g. runtime can create and call an sandboxed
-	// instance which in turn can call the runtime back) we have to be very careful with borrowing
-	// those.
-	//
-	// Basically, most of the interactions should do temporary borrow immediately releasing the
-	// borrow after performing necessary queries/changes.
-	sandbox_store: RefCell<sandbox::Store<SupervisorFuncRef>>,
+	/// We need some interior mutability here since the host state is shared between all host
+	/// function handlers and the wasmtime backend's `impl WasmRuntime`.
+	///
+	/// Furthermore, because of recursive calls (e.g. runtime can create and call an sandboxed
+	/// instance which in turn can call the runtime back) we have to be very careful with borrowing
+	/// those.
+	///
+	/// Basically, most of the interactions should do temporary borrow immediately releasing the
+	/// borrow after performing necessary queries/changes.
+	sandbox_store: Rc<RefCell<sandbox::Store>>,
+	/// The dispatch thunks per sandbox instance.
+	sandbox_dispatch_thunks: RefCell<HashMap<u32, Func>>,
```

Comment:
`dispatch_thunk_per_instance_id`? 

Right now the name is rather vague and there is no indication of what is this `u32` stands for. Alternatively, this could be solved by leaving a comment. The same applies to the wasmi implementation.